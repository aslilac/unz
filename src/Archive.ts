import ArchiveFile from "./ArchiveFile.js";
import ArchiveReader from "./ArchiveReader.js";
import CentralDirectoryListing from "./CentralDirectoryListing.js";
import EndOfCentralDirectory from "./EndOfCentralDirectory.js";

export default class Archive {
	readonly #files: Map<string, ArchiveFile> = new Map();

	constructor(it: ArrayBuffer | ArrayBufferView) {
		const buf =
			it instanceof ArrayBuffer
				? it
				: it.buffer.slice(it.byteOffset, it.byteOffset + it.byteLength);

		const reader = new ArchiveReader(buf);
		const eocd = new EndOfCentralDirectory(reader);
		reader.position = eocd.startOfCentralDirectory;

		for (let i = 0; i < eocd.globalListingCount; i++) {
			const cdl = new CentralDirectoryListing(reader);
			const file = new ArchiveFile(cdl, buf);
			this.#files.set(cdl.fileName, file);
		}

		const size = reader.position - eocd.startOfCentralDirectory;
		if (size !== eocd.sizeOfCentralDirectory) {
			throw new Error("incorrect size for central directory");
		}
	}

	[Symbol.iterator]() {
		return this.#files.entries();
	}

	files() {
		return this.#files.values();
	}

	get(name: string): ArchiveFile | undefined {
		return this.#files.get(name);
	}
}
