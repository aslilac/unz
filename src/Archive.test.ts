import * as fs from "node:fs/promises";
import * as path from "node:path";
import { expect, test } from "vitest";
import Archive from "./Archive.js";

const exampleArchives = [
	"__testdata__/example1.zip",
	"__testdata__/example2.zip",
];

test.each(exampleArchives)(
	"Real archives are parsed properly (%s)",
	async (example) => {
		const buffer = await fs.readFile(path.join(__dirname, example));
		const archive = new Archive(buffer);
		const names = [...archive.files()].map((it) => it.fileName);
		expect(names).toMatchSnapshot();
		const actualSizes = await Promise.all(
			[...archive.files()].map((it) =>
				it.arrayBuffer().then((it) => it.byteLength),
			),
		);
		expect(actualSizes).toMatchSnapshot();
		const sizes = [...archive.files()].map((it) => it.byteLength);
		expect(sizes).toEqual(actualSizes);
		const timestamps = [...archive.files()].map((it) =>
			it.lastModified.toLocaleString(),
		);
		expect(timestamps).toMatchSnapshot();
	},
);
