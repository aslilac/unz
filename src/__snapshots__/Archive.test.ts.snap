// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Real archives are parsed properly (__testdata__/example1.zip) 1`] = `
[
  "wallhaven-6q552w.jpg",
  "__MACOSX/._wallhaven-6q552w.jpg",
  "wallhaven-6qd6rw.jpg",
  "__MACOSX/._wallhaven-6qd6rw.jpg",
  "wallhaven-6qq18w.jpg",
  "__MACOSX/._wallhaven-6qq18w.jpg",
]
`;

exports[`Real archives are parsed properly (__testdata__/example1.zip) 2`] = `
[
  163664,
  188,
  554382,
  188,
  1150840,
  188,
]
`;

exports[`Real archives are parsed properly (__testdata__/example1.zip) 3`] = `
[
  2021-05-26T03:21:44.000Z,
  2021-05-26T03:21:44.000Z,
  2019-11-29T22:39:36.000Z,
  2019-11-29T22:39:36.000Z,
  2021-03-05T22:57:34.000Z,
  2021-03-05T22:57:34.000Z,
]
`;

exports[`Real archives are parsed properly (__testdata__/example2.zip) 1`] = `
[
  "unzzz-f16bf07ca9d203b608c01ee581a0eb435843909c/",
  "unzzz-f16bf07ca9d203b608c01ee581a0eb435843909c/.eslintrc.json",
  "unzzz-f16bf07ca9d203b608c01ee581a0eb435843909c/.gitattributes",
  "unzzz-f16bf07ca9d203b608c01ee581a0eb435843909c/.github/",
  "unzzz-f16bf07ca9d203b608c01ee581a0eb435843909c/.github/workflows/",
  "unzzz-f16bf07ca9d203b608c01ee581a0eb435843909c/.github/workflows/main.yml",
  "unzzz-f16bf07ca9d203b608c01ee581a0eb435843909c/.gitignore",
  "unzzz-f16bf07ca9d203b608c01ee581a0eb435843909c/.nowignore",
  "unzzz-f16bf07ca9d203b608c01ee581a0eb435843909c/.npmignore",
  "unzzz-f16bf07ca9d203b608c01ee581a0eb435843909c/CODE_OF_CONDUCT.md",
  "unzzz-f16bf07ca9d203b608c01ee581a0eb435843909c/LICENSE",
  "unzzz-f16bf07ca9d203b608c01ee581a0eb435843909c/README.md",
  "unzzz-f16bf07ca9d203b608c01ee581a0eb435843909c/jest.config.js",
  "unzzz-f16bf07ca9d203b608c01ee581a0eb435843909c/lib/",
  "unzzz-f16bf07ca9d203b608c01ee581a0eb435843909c/lib/base/",
  "unzzz-f16bf07ca9d203b608c01ee581a0eb435843909c/lib/base/crc.test.ts",
  "unzzz-f16bf07ca9d203b608c01ee581a0eb435843909c/lib/base/crc.ts",
  "unzzz-f16bf07ca9d203b608c01ee581a0eb435843909c/lib/base/reader.test.ts",
  "unzzz-f16bf07ca9d203b608c01ee581a0eb435843909c/lib/base/reader.ts",
  "unzzz-f16bf07ca9d203b608c01ee581a0eb435843909c/lib/cdl.ts",
  "unzzz-f16bf07ca9d203b608c01ee581a0eb435843909c/lib/eocd.ts",
  "unzzz-f16bf07ca9d203b608c01ee581a0eb435843909c/lib/lh.ts",
  "unzzz-f16bf07ca9d203b608c01ee581a0eb435843909c/lib/types/",
  "unzzz-f16bf07ca9d203b608c01ee581a0eb435843909c/lib/types/mappable.ts",
  "unzzz-f16bf07ca9d203b608c01ee581a0eb435843909c/lib/types/signatures.ts",
  "unzzz-f16bf07ca9d203b608c01ee581a0eb435843909c/lib/unzzz.test.ts",
  "unzzz-f16bf07ca9d203b608c01ee581a0eb435843909c/lib/unzzz.ts",
  "unzzz-f16bf07ca9d203b608c01ee581a0eb435843909c/now.json",
  "unzzz-f16bf07ca9d203b608c01ee581a0eb435843909c/package.json",
  "unzzz-f16bf07ca9d203b608c01ee581a0eb435843909c/prettier.config.js",
  "unzzz-f16bf07ca9d203b608c01ee581a0eb435843909c/tsconfig.json",
  "unzzz-f16bf07ca9d203b608c01ee581a0eb435843909c/yarn.lock",
]
`;

exports[`Real archives are parsed properly (__testdata__/example2.zip) 2`] = `
[
  0,
  173,
  73,
  0,
  0,
  650,
  102,
  102,
  37,
  114,
  1073,
  1754,
  116,
  0,
  0,
  252,
  451,
  3057,
  1474,
  3622,
  1487,
  3044,
  0,
  70,
  159,
  3319,
  7532,
  226,
  877,
  169,
  382,
  200472,
]
`;

exports[`Real archives are parsed properly (__testdata__/example2.zip) 3`] = `
[
  2020-12-24T15:40:50.000Z,
  2020-12-24T15:40:50.000Z,
  2020-12-24T15:40:50.000Z,
  2020-12-24T15:40:50.000Z,
  2020-12-24T15:40:50.000Z,
  2020-12-24T15:40:50.000Z,
  2020-12-24T15:40:50.000Z,
  2020-12-24T15:40:50.000Z,
  2020-12-24T15:40:50.000Z,
  2020-12-24T15:40:50.000Z,
  2020-12-24T15:40:50.000Z,
  2020-12-24T15:40:50.000Z,
  2020-12-24T15:40:50.000Z,
  2020-12-24T15:40:50.000Z,
  2020-12-24T15:40:50.000Z,
  2020-12-24T15:40:50.000Z,
  2020-12-24T15:40:50.000Z,
  2020-12-24T15:40:50.000Z,
  2020-12-24T15:40:50.000Z,
  2020-12-24T15:40:50.000Z,
  2020-12-24T15:40:50.000Z,
  2020-12-24T15:40:50.000Z,
  2020-12-24T15:40:50.000Z,
  2020-12-24T15:40:50.000Z,
  2020-12-24T15:40:50.000Z,
  2020-12-24T15:40:50.000Z,
  2020-12-24T15:40:50.000Z,
  2020-12-24T15:40:50.000Z,
  2020-12-24T15:40:50.000Z,
  2020-12-24T15:40:50.000Z,
  2020-12-24T15:40:50.000Z,
  2020-12-24T15:40:50.000Z,
]
`;
